firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/FirestoreMapDecoder.kt

```diff
@@ -0,0 +1,244 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.ktx.serialization
+
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.ktx.serialization.FirestoreAbstractDecoder.Constants.START_INDEX
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.SerialKind
+import kotlinx.serialization.descriptors.StructureKind
+import kotlinx.serialization.descriptors.elementNames
+import kotlinx.serialization.encoding.AbstractDecoder
+import kotlinx.serialization.encoding.CompositeDecoder
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.modules.EmptySerializersModule
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.serializer
+
+/**
+ * A skeleton implementation of a decoder that decodes a nested map (or list) of Firestore supported
+ * types as a @[Serializable] object (or its field).
+ *
+ * Most of the decode* methods have default implementation that delegates decodeValue(value: Any).
+ * The [decodeElementIndex] method need to be override in each of the subclasses to support decoding
+ * of a nested map or list.
+ *
+ * @param nestedObject The nested object (map or list) that that needs to be decoded.
+ * @param docRef The [DocumentReference] where this nested object is obtained from.
+ */
+private abstract class FirestoreAbstractDecoder(
+    private val nestedObject: Any,
+    private val docRef: DocumentReference
+) : AbstractDecoder() {
+
+    object Constants {
+        const val START_INDEX = 0
```

Comment:
Not sure this `const` enhances readability.

Reply:
Very interesting, the `const` keyword helps prevent the generated getter in the Java decomplied code:

1. This is the current code's decomplied Java code:
```
public static final int START_INDEX = 0;
```
2. If I remove the `const` keyword, there will be a getter generated for it like a property:
```
      private static final int START_INDEX;
      @NotNull
      public static final FirestoreAbstractDecoder.Constants INSTANCE;

      public final int getSTART_INDEX() {
         return START_INDEX;
      }
      ```

Reply:
I think we do not need a getter here, I think Kotlin is recommend us to use `const` whenever we can: 
The advantages of keep this number as a compile time `const` are:
1. the decompiled Java code is shorter,
2. this `const` will be inlined (directly use the primitive value to replace the variable in the compiled code), my understanding is this can reduce the runtime heap